scanpy/external/pp/_hashsolo.py

```diff
@@ -0,0 +1,387 @@
+        )
+        probs_of_singlet = np.sum(
+            [log_noise_noise_probs, log_signal_signal_probs], axis=0
+        )
+        probs_of_doublet = np.sum(
+            [log_noise_signal_probs, log_signal_signal_probs], axis=0
+        )
+        log_probs_list = [probs_of_negative, probs_of_singlet, probs_of_doublet]
+
+        # each cell and each hypothesis probability
+        for prob_idx, log_prob in enumerate(log_probs_list):
+            log_likelihoods_for_each_hypothesis[indices, prob_idx] = log_prob
+    return log_likelihoods_for_each_hypothesis, all_indices, counter_to_barcode_combo
+
+
+def _calculate_bayes_rule(data, priors, number_of_noise_barcodes):
+    """
+    Calculate bayes rule from log likelihoods
+    Parameters
+    ----------
+    data : np.array
+        Anndata object filled only with hashing counts
+    priors : list,
+        a list of your prior for each hypothesis
+        first element is your prior for the negative hypothesis
+        second element is your prior for the singlet hypothesis
+        third element is your prior for the doublet hypothesis
+        We use [0.01, 0.8, 0.19] by default because we assume the barcodes
+        in your cell hashing matrix are those cells which have passed QC
+        in the transcriptome space, e.g. UMI counts, pct mito reads, etc.
+    number_of_noise_barcodes : int
+        number of barcodes to used to calculated noise distribution
+    Returns
+    -------
+    bayes_dict_results : dict
+        "most_likely_hypothesis" key is a 1d np.array of the most likely hypothesis
+        "probs_hypotheses" key is a 2d np.array probability of each hypothesis
+        "log_likelihoods_for_each_hypothesis" key is a 2d np.array log likelihood of each hypothesis
+    """
+    priors = np.array(priors)
+    log_likelihoods_for_each_hypothesis, _, _ = _calculate_log_likelihoods(
+        data, number_of_noise_barcodes
+    )
+    probs_hypotheses = (
+        np.exp(log_likelihoods_for_each_hypothesis)
+        * priors
+        / np.sum(
+            np.multiply(np.exp(log_likelihoods_for_each_hypothesis), priors), axis=1
+        )[:, None]
+    )
+    most_likely_hypothesis = np.argmax(probs_hypotheses, axis=1)
+    return {
+        "most_likely_hypothesis": most_likely_hypothesis,
+        "probs_hypotheses": probs_hypotheses,
+        "log_likelihoods_for_each_hypothesis": log_likelihoods_for_each_hypothesis,
+    }
+
+
+def hashsolo(
+    cell_hashing_adata: anndata.AnnData,
+    cell_hashing_columns: list,
+    priors: list = [0.01, 0.8, 0.19],
+    pre_existing_clusters: str = None,
+    number_of_noise_barcodes: int = None,
+    inplace: bool = True,
+):
+    """Probabilistic demultiplexing of cell hashing data using HashSolo [Bernstein20]_.
+
+    .. note::
+        More information and bug reports `here <https://github.com/calico/solo>`__.
+    Parameters
+    ----------
+    cell_hashing_adata : anndata.AnnData
```

Comment:
```suggestion
    adata : anndata.AnnData
```

It's a bit confusing to me to call it `cell_hashing_adata` as it makes it seem the anndata should only have cell hashing data in it

Reply:
Ya that makes sense. This is a remnant of me porting the code from the solo repo which expects an `anndata` with hashing counts in `.X`

Reply:
done